File: openharmony/arkcompiler_runtime_core/static_core/plugins/ets/tests/interop_js/tests/indexed_type/indexed_type.js
'use strict';
let __extends =
	(this && this.__extends) ||
	(function () {
		let extendStatics = function (d, b) {
			extendStatics =
				Object.setPrototypeOf ||
				({ __proto__: [] } instanceof Array &&
					function (d, b) {
						d.__proto__ = b;
					}) ||
				function (d, b) {
					for (let p in b) {
						if (Object.prototype.hasOwnProperty.call(b, p)) {
							d[p] = b[p];
						}
					}
				};
			return extendStatics(d, b);
		};
		return function (d, b) {
			if (typeof b !== 'function' && b !== null) {
				throw new TypeError('Class extends value ' + String(b) + ' is not a constructor or null');
			}
			extendStatics(d, b);
			function Ctor() {
				this.constructor = d;
			}
			d.prototype = b === null ? Object.create(b) : ((Ctor.prototype = b.prototype), new Ctor());
		};
	})();
exports.__esModule = true;
exports.handler = exports.createInt8Array = exports.createStringRecord = exports.returnCustomArray = exports.CustomArray = exports.getArrey = void 0;
/**
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
function getArray(arr) {
	return arr;
}
exports.getArray = getArray;
let customArray = /** @class */ (function (_super) {
	__extends(customArray, _super);
	function customArray(...args) {
		let items = [];
		for (let _i = 0; _i < args.length; _i++) {
			items[_i] = args[_i];
		}
		return _super.apply(this, items) || this;
	}
	return customArray;
})(Array);
exports.CustomArray = customArray;
function returnCustomArray(array) {
	return array;
}
exports.returnCustomArray = returnCustomArray;
function createStringRecord() {
	let data = {
		key1: 'A',
		key2: 'B',
		key3: 'C',
	};
	return data;
}
exports.createStringRecord = createStringRecord;
function createInt8Array(length) {
	let array = new Int8Array(length);
	for (let i = 0; i < length; i++) {
		array[i] = i;
	}
	return array;
}
exports.createInt8Array = createInt8Array;
exports.handler = {
	get: function (target, property) {
		if (property in target) {
			return target[property];
		} else {
			return undefined;
		}
	},
	set: function (target, property, value) {
		if (typeof value === 'string') {
			target[property] = value;
			return true;
		} else {
			throw new TypeError('The value must be a string');
		}
	},
};

File: openharmony/arkcompiler_runtime_core/static_core/plugins/ets/tests/interop_js/tests/indexed_type/indexed_type.cpp
/**
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <gtest/gtest.h>
#include "ets_interop_js_gtest.h"

namespace ark::ets::interop::js::testing {

class EtsIndexedType : public EtsInteropTest {};

TEST_F(EtsIndexedType, getArrValueByIndex)
{
    auto ret = CallEtsMethod<bool>("getArrValueByIndex");
    ASSERT_EQ(ret, true);
}
TEST_F(EtsIndexedType, changeArrValueByIndex)
{
    auto ret = CallEtsMethod<bool>("changeArrValueByIndex");
    ASSERT_EQ(ret, true);
}
TEST_F(EtsIndexedType, checkLengthArr)
{
    auto ret = CallEtsMethod<bool>("checkLengthArr");
    ASSERT_EQ(ret, true);
}
TEST_F(EtsIndexedType, checkAllArrValue)
{
    auto ret = CallEtsMethod<bool>("checkAllArrValue");
    ASSERT_EQ(ret, true);
}
TEST_F(EtsIndexedType, getCustomArrValueByIndex)
{
    auto ret = CallEtsMethod<bool>("getCustomArrValueByIndex");
    ASSERT_EQ(ret, true);
}
TEST_F(EtsIndexedType, changeCustomArrValueByIndex)
{
    auto ret = CallEtsMethod<bool>("changeCustomArrValueByIndex");
    ASSERT_EQ(ret, true);
}
TEST_F(EtsIndexedType, checkLengthCustomArr)
{
    auto ret = CallEtsMethod<bool>("checkLengthCustomArr");
    ASSERT_EQ(ret, true);
}
TEST_F(EtsIndexedType, checkAllCustomArrValue)
{
    auto ret = CallEtsMethod<bool>("checkAllCustomArrValue");
    ASSERT_EQ(ret, true);
}
TEST_F(EtsIndexedType, getRecordValue)
{
    auto ret = CallEtsMethod<bool>("getRecordValue");
    ASSERT_EQ(ret, true);
}
TEST_F(EtsIndexedType, changeRecordValue)
{
    auto ret = CallEtsMethod<bool>("changeRecordValue");
    ASSERT_EQ(ret, true);
}
// NOTE(andreypetukhov) enable after fixibng #17821
TEST_F(EtsIndexedType, DISABLED_getRecordValueIndex)
{
    auto ret = CallEtsMethod<bool>("getRecordValueIndex");
    ASSERT_EQ(ret, true);
}
// NOTE(andreypetukhov) enable after fixibng #17821
TEST_F(EtsIndexedType, DISABLED_changeRecordValueIndex)
{
    auto ret = CallEtsMethod<bool>("changrRecordValueIndex");
    ASSERT_EQ(ret, true);
}
// NOTE(andreypetukhov) enable after fixibng #17821
TEST_F(EtsIndexedType, DISABLED_checkRecordValueLength)
{
    auto ret = CallEtsMethod<bool>("checkRecordValueLength");
    ASSERT_EQ(ret, true);
}
// NOTE(andreypetukhov) enable after fixibng #17821
TEST_F(EtsIndexedType, DISABLED_checkRecordValue)
{
    auto ret = CallEtsMethod<bool>("checkRecordValue");
    ASSERT_EQ(ret, true);
}
TEST_F(EtsIndexedType, getTypedArrayValueByIndex)
{
    auto ret = CallEtsMethod<bool>("getTypedArrayValueByIndex");
    ASSERT_EQ(ret, true);
}
// NOTE(andreypetukhov) enable after fixibng #18131
TEST_F(EtsIndexedType, DISABLED_changeTypedArrayValueByIndex)
{
    auto ret = CallEtsMethod<bool>("changeTypedArrayValueByIndex");
    ASSERT_EQ(ret, true);
}
// NOTE(andreypetukhov) enable after fixibng #18131
TEST_F(EtsIndexedType, DISABLED_checkLengthTypedArray)
{
    auto ret = CallEtsMethod<bool>("checkLengthTypedArray");
    ASSERT_EQ(ret, true);
}
// NOTE(andreypetukhov) enable after fixibng #18131
TEST_F(EtsIndexedType, DISABLED_checkAllTypedArrayValue)
{
    auto ret = CallEtsMethod<bool>("checkAllTypedArrayValue");
    ASSERT_EQ(ret, true);
}
TEST_F(EtsIndexedType, getValueFromProxyRecord)
{
    auto ret = CallEtsMethod<bool>("getValueFromProxyRecord");
    ASSERT_EQ(ret, true);
}
TEST_F(EtsIndexedType, changeValueFromProxyRecord)
{
    auto ret = CallEtsMethod<bool>("changeValueFromProxyRecord");
    ASSERT_EQ(ret, true);
}
// NOTE(andreypetukhov) enable after fixibng #18238
TEST_F(EtsIndexedType, DISABLED_getValueFromProxyByKeyRecord)
{
    auto ret = CallEtsMethod<bool>("getValueFromProxyByKeyRecord");
    ASSERT_EQ(ret, true);
}

}  // namespace ark::ets::interop::js::testing